Sources/XiCliCore/CommandLineTool.swift

```diff
@@ -0,0 +1,155 @@
+// Copyright 2018 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Cocoa
+
+public final class CommandLineTool {
+    private let arguments: [String]
+    
+    public init(arguments: [String] = CommandLine.arguments) {
+        self.arguments = Array(arguments.dropFirst())
```

Comment:
An architectural thought:

Valid arguments are a precondition for continued execution. With that in mind, if I were writing this utility, I would approach it slightly differently:

- I would move `getArguments` out of `CommandLineTool`, and make it a constructor on `Arguments`. Then the very first thing I would do when the application launches is to try and parse the raw arguments into an `Arguments` struct. If this failed, for any reason, I would print the help message and abort, before even `init`ing `CommandLineTool`.
- I would make `CommandLineTool` take an `Arguments` struct as it's constructor. At this point, we _know_ that the input is well formed, and the tool class only deals with known inputs.

This makes our testing clearer, and is generally (I think) a clearer separation of responsibilities. `Arguments` handles parsing raw arguments, and `CommandLineTool` only has to deal with the actual logic of executing the provided arguments.

In any case, I'm just offering this up as an observation of how _I_ would approach this problem; I think the current version is okay, but I thought sharing that might be helpful.

edit: the one real complaint I would have about this function is that if you want to actually pass it arguments directly, you're expected to include an empty first argument. That feels weird.

Reply:
I have the refactor almost done. As for the edit, the command line treats the `xi` in `xi <filepath> [flags]` as the first argument, not removing that would cause the CLI to open a file named `xi` whenever no additional arguments are provided from the command line.

Reply:
right, I'm not saying that we shouldn't be dropping the first argument, just that it feels like that is part of the process of parsing raw args, and should be happening when we do that, not in the initializer for this class. 👍

Reply:
Oh, ok. In the current refactor I moved it to the initializer for the Argument struct, do you think that's better?

Reply:
Yea, I think that makes more sense.